File: benches/class_dispatch/class_dispatch.go
package main

// Count from stdin so the loop can't be constant-folded; `inc` is
// //go:noinline so each iteration is a genuine call, matching PHC's
// @noinline (D-052) and the C/Rust noinline builds.
import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Counter struct{ value int64 }

//go:noinline
func (c *Counter) inc() { c.value++ }

func (c *Counter) get() int64 { return c.value }

func main() {
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	count, _ := strconv.ParseInt(strings.TrimSpace(line), 10, 64)
	c := &Counter{value: 0}
	var i int64 = 0
	for i < count {
		c.inc()
		i++
	}
	fmt.Printf("counter = %d\n", c.get())
}

File: benches/class_dispatch/class_dispatch.rs
// Count from stdin so the loop can't be constant-folded; `inc` is
// `#[inline(never)]` so each iteration is a genuine call, matching
// PHC's @noinline (D-052) and the C noinline build.
use std::io::Read;

struct Counter {
    value: i64,
}

impl Counter {
    fn new(start: i64) -> Counter {
        Counter { value: start }
    }
    #[inline(never)]
    fn inc(&mut self) {
        self.value += 1;
    }
    fn get(&self) -> i64 {
        self.value
    }
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_to_string(&mut input).unwrap();
    let count: i64 = input.trim().parse().unwrap_or(0);
    let mut c = Counter::new(0);
    let mut i: i64 = 0;
    while i < count {
        c.inc();
        i += 1;
    }
    println!("counter = {}", c.get());
}
